### PR TITLE
[WIP] Multi channel support & more advanced permissions.

### PR DESCRIPTION
**Pull request by lamafab**
_Work-in-Progress, do NOT merge yet_

This PR fixes the following two issues:

* https://github.com/w3f/SecOps/issues/91
* https://github.com/w3f/SecOps/issues/92

The current codebase makes a lot of assumptions about the usage of the service, respectively that the primary way to interact with it and to modify state is via the Matrix interface. Since we want to support multiple channels and therefore multiple ways to escalate and acknowledge escalations, I decided to do some refactoring. A big change in this PR is that it moves most of the logic away from the Matrix "adapter" into a separate, generic task that handles escalations and acknowledgement independently from the adapter (Matrix, PagerDuty, Email, etc). This allows us to keep any adapter quite thin and to reuse code. And most importantly: it's easier to (unit) test.

Task list:

* [x] Generic and independent handling of escalations.
* [x] Generic and independent handling of acknowledgements.
* [x] Implement PagerDuty adapter (after refactor).
  * [x] Support acknowledgements over PagerDuty. 
* [x] Strip current Matrix adapter of core logic functionality, unify with all adapters.
* [x] Implement email adapter.
* [x] More fine-grained control of roles/permission (see [sample config](https://github.com/w3f/matrixbot-ack/blob/8a09c93834451685f796a4fdf37f78dbd5758f15/config.sample.yaml))
* [x] Update database schema and interactions.
* [x] (Mocked) unit tests of core logic and behavior.

### CODE REVIEW

File: charts/matrixbot-ack/values.yaml [hunk 1]
```diff
@@ -16,10 +16,15 @@ config:
     db_path: /app/db/matrix.db
     device_name: matrixbot-ack
     device_id: matrixbot-some-id
+  pager_duty:
```
Comment (ironoa) [inline charts/matrixbot-ack/values.yaml:19]: can we move `pager_duty`to `escalation.pager_duty` ?

File: charts/matrixbot-ack/values.yaml [hunk 2]
```diff
@@ -16,10 +16,15 @@ config:
...
+    api_key: abcdefg
```
Comment (ironoa) [inline charts/matrixbot-ack/values.yaml:20]: can we add an `pager_duty.enabled`value ? default to `false`

File: config.sample.yaml [hunk 3]
```diff
@@ -1,17 +1,39 @@
...
+    - "!abcdef:matrix.org"
+    - "!ghijkl:matrix.org"
+    config:
+      homeserver: https://matrix.org
+      username: username
+      password: password
+      db_path: db/matrix.db
+      device_name: matrixbot-ack
+      device_id: matrixbot-some-id
+  pager_duty:
```
Comment (ironoa) [inline config.sample.yaml:20]: same: can we move pager_duty to escalation.pager_duty ?

File: helmfile.d/config/matrixbot-alert.yaml.gotmpl [hunk 4]
```diff
@@ -18,9 +18,14 @@ config:
     device_name: {{ env "MATRIX_DEVICE_NAME" | quote }}
     device_id: {{ env "MATRIX_DEVICE_ID" | quote }}
     db_path: db/matrix.db
+  pager_duty:
```
Comment (ironoa) [inline helmfile.d/config/matrixbot-alert.yaml.gotmpl:21]: same: can we move pager_duty to escalation.pager_duty ? and add an enabed value ?

File: helmfile.d/config/matrixbot-escalation.yaml.gotmpl [hunk 5]
```diff
@@ -20,10 +20,15 @@ config:
     device_name: {{ env "MATRIX_DEVICE_NAME" | quote }}
     device_id: {{ env "MATRIX_DEVICE_ID" | quote }}
     db_path: db/matrix.db
+  pager_duty:
```
Comment (ironoa) [inline helmfile.d/config/matrixbot-escalation.yaml.gotmpl:23]: same: can we move pager_duty to escalation.pager_duty ? and add an enabed value ?

File: charts/matrixbot-ack/values.yaml (hunk 2 above)
Comment (lamafab) [reply charts/matrixbot-ack/values.yaml:20]: It does have that field, see https://github.com/w3f/matrixbot-ack/blob/multi-channel-restruct/config.sample.yaml#L21

File: charts/matrixbot-ack/values.yaml (hunk 1 above)
Comment (lamafab) [reply charts/matrixbot-ack/values.yaml:19]: So, from the perspective of the service, pagerduty is just another "adapter". Meaning that matrix and pagerduty are treated the same and pretty much behave the same, although pagerduty now has that `pager_duty.only_on_escalation` field. IMO we should leave it like that.

Comment (lamafab) [reply charts/matrixbot-ack/values.yaml:19]: Yeah that `charts/matrixbot-ack/values.yaml` hasn't been updated in a while it seems...

**Comment by lamafab** [comment]
We decided (months ago) to deprecate this. Might still be useful to keep this branch around.

**Pull request closed without merging**